### DISCUSSION TITLE
INC3 common module design collections

### DISCUSSION THREAD

**Discussion by yiliu30** [intel/neural-compressor#1445] [RFC]
> Note: INC3  is currently under heavy development, APIs are subject to change.

### Goal
- Enable users to easily utilize the predefined tuning space and tuning order with just a few lines of code.
- Allow users to tune a single algorithm or combine multiple algorithms.
- Empower developers to register framework-specific tuning space and tuning order.
- Provide developers with the flexibility to customize the framework-specific autotune pipeline according to their needs.

### Design Overview
- Maintain a thin autotune layer that offers the common components to drive the tuning pipeline while allowing framework developers to customize it. These components include:
    - `TuningConfig`: Used by users to set tuning spaces, tuning order, and stop conditions.
    -  `TuningLogger`: Record the tuning process log, facilitating the automatic collection of tuning results by validation teams.
    - `ConfigLoader`: Takes the config set and sampler, yielding quantization configuration one by one.
    - `TuningMonitor`: Records trial information and provides interfaces to check stop conditions.
    - `Evaluator`: Wraps user-provided evaluation functions into a unified interface to obtain the final evaluation score.

### The `autotune` API
One of the major changes in INC 3 is the separation of quantization and autotune into two distinct APIs. Quantization APIs align with stock frameworks.  The `autotune` is the tuning interface used by all framework extensions, with framework-specific arguments.  The `autotune` API accepts two common arguments: tune_config and eval_fns.
```python
def autotune(
    model,
    tune_config: TuningConfig,
    eval_fns: Optional[Union[Callable, Dict, List[Dict]]] = None,
    ...
) -> Optional[torch.nn.Module]:
    """The main entry of auto-tune."""

```

### Usage Examples
```python
# User scripts
from typing import Union
import torch

class UserModel(torch.nn.Module):
    def __init__(self, dims):
        self.fc1 = torch.nn.Linear(dims, dims)
        self.fc2 = torch.nn.Linear(dims, dims)
        self.fc3 = torch.nn.Linear(dims, dims)

    def forward(self, x):
        x = self.fc1(x)
        x = self.fc2(x)
        x = self.fc3(x)
        return x


def build_simple_model(dims=10):
    return UserModel(dims)


def eval_fn(model) -> Union[float, int]:
    # Define a function to evaluate the model and return a score(float or int).
    # ...
    return 1.0


def calib_fn(model):
    #  Define a function to calibrate the model.
    # ...
    pass


# 1. Use the default tuning space.
from neural_compressor.torch.quantization import GPTQConfig, RTNConfig, TuningConfig, autotune

float_model = build_simple_model()
rtn_default_config_set = RTNConfig.get_config_set_for_tuning()
q_model = autotune(model=float_model, tune_config=TuningConfig(rtn_default_config_set), eval_fns=eval_fn)

# 2.1 Customize the tuning space for a single algorithm.
float_model = build_simple_model()
customize_config_set = TuningConfig(config_set=RTNConfig(bits=[4, 8]))
q_model = autotune(model=float_model, tune_config=TuningConfig(customize_config_set), eval_fns=eval_fn)

# 2.2 Combine multiple algorithms into the tuning space.
float_model = build_simple_model()
customize_config_set = [RTNConfig(bits=[4, 8]), GPTQConfig(bits=[4, 8])]
q_model = autotune(model=float_model, tune_config=TuningConfig(customize_config_set), run_fn=calib_fn, eval_fns=eval_fn)
```

## Advantage Topics [WIP]
- Define multiple evaluation functions

- The relationship between `ConfigLoader`, `Sampler`, `ConfigSet`, and `XxxAgloConfig`.

- Customize the tuning order

-- END

**Comment by yiliu30** [comment]
Notes

- Integration Example:
  - [ ] With DeepSpeed
- [ ] Consider TP/DP/PP for eval/calibrtate/prepare/quantize
- [x] Refine the naming of `Runner`
- [x] Support register different objectives
- [x] Decompose the quantize into calibrtate/prepare/quantize

**Comment by yiliu30** [comment]
### Task List

1) 1st PR Introduce the overview of the auto-tune pipeline and main interface.   https://github.com/intel/neural-compressor/pull/1497
- [x] Main components and their declaration, relationship: `Tuner`, `TaskManager`, `FrameworkWrapper`, and `TorchWrapper`
- [x] Tune-related config `BaseTuningConfig`, `TuneConfig`

2) 2nd PR
- [ ] Merge user config with framework registered config
- [x] Expand the config with local
- [x] Register the default tuning order, which needs input from fwk owners

3) 3rd PR
- [ ] Docs

**Comment by yiliu30** [comment]
## About the tuning space and tuning order.

### Case 1. Use the default tuning space and tuning order

```python
# for rtn
default_rtn_tuning_config = TuningConfig(quant_configs=get_default_rtn_quant_configs(), max_trials=100)
```

### Case 2. The user specifies the tuning space and tuning order

> The concept of `sampler`: a `sampler` takes a set of quant configs and yields the config for one quantization process with a specific order.

1) User customize the tuning space(use the default `sampler` to sample the quant_config)

```python
customized_rtn_configs = RTNWeightQuantConfig(weight_bits=[2, 4, 6, 8])
tuning_config = TuningConfig(quant_configs=customized_rtn_configs, max_trials=100)
```

2) User customizes both tuning space and tuning order (`sampler`)

```python
def customized_sampler(config: RTNWeightQuantConfig) -> List[RTNWeightQuantConfig]:
    ...

customized_rtn_configs = RTNWeightQuantConfig(weight_bits=[2, 4, 6, 8])
customized_rtn_configs.set_sampler(customized_sampler)
tuning_config = TuningConfig(quant_configs=customized_rtn_configs, max_trials=100)
```

**Reply by yiliu30** [reply]
One item that may confuse users is that:
 - `RTNWeightQuantConfig(weight_bits=2)` used by the `quantize` API represents one trial
 - `RTNWeightQuantConfig(weight_bits=[2, 4, 6, 8])` used by the `autotune` API represents four trials

**Reply by yiliu30** [reply]
@ftian1 please have a look, thank you.

**Reply by yiliu30** [reply]
After syncing with @ftian1 offline, here is the summary:

1. Keep using one class `RTNWeightQuantConfig` in both `quantize` and `autotune` APIs.
```python
RTNWeightQuantConfig(weight_bits=2) #used by the quantize API represents one trial
RTNWeightQuantConfig(weight_bits=[2, 4, 6, 8]) #used by the autotune API represents four trials
```
2. The agrs of `TuningConfig`:

- Add `sampler` for sampling config from space
- Rename `quant_configs` to `tuning_space`
 
```python
tuning_space = get_default_tuning_space() 
# return all built-in spaces  {rtn_tuning_space, gptq_tuning_space, sq_tuning_space, static_tuning_space}
tuning_config = TuningConfig(tuning_space=tuning_space, max_trials=100, sampler:Optional[Sampler])
```

**Comment by yiliu30** [comment]
### ConfigRegistry

```python

@register_config(framework_name=FRAMEWORK_NAME, algo_name=GPTQ, priority=100)

#registered_configs
class ConfigRegistry:
    
    
FRAMEWORK_NAME:
    ALGORITHM_NAME:
        PRIORITY
        CLS
        
config_registry = ConfigRegistry

config_registry.get_all_configs()
config_registry.get_sorted_configs()


1. add `priority` into `register_config`
2. replace `registered_configs` with `config_registry`

```

**Reply by yiliu30** [reply]
@Kaihui-intel , please add the PR link here.

**Reply by Kaihui-intel** [reply]
https://github.com/intel/neural-compressor/pull/1543

**Comment by yiliu30** [comment]
## About the implementation details of `Tuning`

```python

def autotune(...):
    ....
    #####  Option 1.  Original Impl
    tuning_objectives.set_eval_fn_registry(eval_fns)
    torch_wrapper = TorchWrapper(model, run_fn, run_args)
    tuner = Tuner(tune_config=tune_config, tuning_objectives=tuning_objectives, fwk_wrapper=torch_wrapper)
    best_qmodel = tuner.search()
    return  best_qmodel
    
    #####  Option 2. New impl
    from neural_compressor.torch import quantize
    from nueral_compressor.common import init_tuning
    sampler, evaluator, recorder = init_tuning(tune_config)
    for quant_config: BaseConfig in sampler.next_config():
        q_model = quantize(model=model, quant_config=quant_config, run_fn=run_fn, run_args=run_args)
        eval_result = evaluator.eval(q_model)
        recorder.add_new_trial(quant_config, eval_result)
        if recorder.needs_stop():
            best_qmodel = recorder.get_best_qmodel()
            return best_qmodel
    return best_qmodel
```

**Reply by yiliu30** [reply]
|                         | Option1                                                      | Option2                                                |
| ----------------------- | ------------------------------------------------------------ | ------------------------------------------------------ |
| Tuning code reusability | **Good**<br />Fwks create a `Tuner` and transfer the ownership to `Tuner`. | **Bad**<br />Fwks needs implement the pipeline.        |
| Custom tuning pipeline  | **No easy** <br />The pipeline is common across fwks         | **Easy** <br />Fwks owns the pipeline and can customize it.                 |
| Call flow               | **Complex**<br /> `torch` call functions from `common`, <br />`common` call functions from `torch`(by wrapper). | **Simple**<br /> `torch` call functions from `common`. |

**Reply by yiliu30** [reply]
Need your comments @xin3he @mengniwang95 @zehao-intel , thanks :).

**Reply by yiliu30** [reply]
Option 1 (React Here)

**Reply by yiliu30** [reply]
Option 2 (React Here)

**Reply by yiliu30** [reply]
After offline syncing, Option 2 was selected, and PR will be added soon.

**Reply by yiliu30** [reply]
https://github.com/intel/neural-compressor/pull/1528

**Comment by yiliu30** [comment]
- [x] Flag indicating the end of quantization(For CI)

![MicrosoftTeams-image (2)](https://github.com/intel/neural-compressor/assets/106061964/2ec31da5-b291-47c6-b891-db0b3896a4dd)

**Reply by yiliu30** [reply]
https://github.com/intel/neural-compressor/pull/1528

**Comment by yiliu30** [comment]
### Workspace
- [x] Add workspace at common
We should port the `workspace` in 2.x into 3.x common.

**Reply by yiliu30** [reply]
@Kaihui-intel please have a look.

**Reply by Kaihui-intel** [reply]
https://github.com/intel/neural-compressor/pull/1553

**Comment by yiliu30** [comment]
### Enhance the `TuningConfig`

- [ ] Add into `tolerable_loss` into `TuningConfig`

Currently, we only support stopping the tuning process when `max_Trials` is reached. Add support for stopping when the accuracy target is reached.

https://github.com/intel/neural-compressor/blob/9ec22c96993d1a8dc00e0cebca1cbe5527ea8189/neural_compressor/config.py#L502-L511

**Reply by yiliu30** [reply]
Depends on https://github.com/intel/neural-compressor/pull/1528

**Reply by yiliu30** [reply]
To clarify, we can only add `tolerable_loss ` as the other two args can be covered by it.

**Reply by yiliu30** [reply]
```python
class TuningConfig:
    """Base Class for Tuning Criterion.
    def __init__(self, config_set=None, timeout=0, max_trials=100, sampler: Sampler = None, tolerable_loss = 0.01) -> None:
        """Init a TuneCriterion object."""
        self.config_set = config_set
        self.timeout = timeout
        self.max_trials = max_trials
        self.sampler = sampler
        self.tolerable_loss = tolerable_loss  # <-- New added
```

**Reply by Kaihui-intel** [reply]
https://github.com/intel/neural-compressor/pull/1579

**Comment by yiliu30** [comment]
### Default tuning space
- [x] Enable the `default_tuning_space`

**Reply by yiliu30** [reply]
https://github.com/intel/neural-compressor/pull/1562

**Comment by yiliu30** [comment]
Tasks List https://github.com/orgs/intel/projects/54/views/1?filterQuery=

**Comment by yiliu30** [comment]
### Divide the `params_list` into model-level, op-type-level, op-level
- [ ] Need a solution

**Comment by yiliu30** [comment]
### Idea about simplifying the CI for common components.

```bash


# torch/3x/torch/test_common.py
Tests for common components.
Owner(s): ["module: common & auto-tune"]

These tests aim to assess the fundamental functionalities of common components and enhance code coverage. 
Currently, there are three replicas in each framework test folder. We may organize them into individual folders
like 'test/3x/common' and update the CI scripts to include them in each framework's CI.

The folder structure:
.
├── 3x
│   ├── common  # <---- New added
│   ├── onnxrt
│   ├── tensorflow
│   └── torch

For each fwk CI:

onnxrt_included_folder:
    ├── 3x
    │   ├── common
    │   ├── onnxrt
    
tensorflow_included_folder:
    ├── 3x
    │   ├── common
    │   ├── tensorflow


torch_included_folder:
    ├── 3x
    │   ├── common
    │   ├── torch

```

**Reply by yiliu30** [reply]
@chensuyue @XuehaoSun Please have a look. Thanks :)

**Reply by chensuyue** [reply]
I can update the CI test logical when the new structure submitted.

**Reply by yiliu30** [reply]
https://github.com/intel/neural-compressor/pull/1575

**Comment by yiliu30** [comment]
## Folder structure of INC 3.X

### !!! Avoid creating a folder for just a single file !!!

```bash
├── fwk_name
│   ├── __init__.py
│   ├── quantization
│   │   ├── algorithm_entry.py
│   │   ├── autotune.py
│   │   ├── config.py
│   │   ├── __init__.py
│   │   └── quantize.py
│   ├── algorithms
│   │   ├── __init__.py
│   │   ├── smooth_quant
│   │   │   ├── __init__.py
│   │   │   ├── smooth_quant.py
│   │   │   └── utility.py
│   │   ├── static_quant
│   │   │   ├── __init__.py
│   │   │   ├── static_quant.py
│   │   │   └── utility.py
│   │   └── weight_only
│   │       ├── gptq.py
│   │       ├── __init__.py
│   │       └── rtn.py
│   └── utils
│       ├── constants.py
│       ├── __init__.py
│       └── utility.py
└── __init__.py
```

``` python
# * Note some code snippets
# neural_compressor/fwk_name/quantization/algorithm_entry.py
@register_algo(RTN)
def rtn_algo_entry()
    from neural_compressor.fwk_name.algorithms import rtn
    ...

@register_algo(SMOOTH_QUANT)
def smooth_quant_entry():
    from neural_compressor.fwk_name.algorithms import smooth_quant
    ...
    
```

**Reply by yiliu30** [reply]
@ftian1 @mengniwang95 @xin3he @zehao-intel

**Comment by Kaihui-intel** [comment]
### Enhance autotune UT to replace checking logs

**Comment by yiliu30** [comment]
### Separate Smooth and Quant from smooth quantization [WIP]

```python
# Option 1 (Currently implementation)
from neural_compressor.torch.quantization import SmoothQuantConfig, quantize
sq_config = SmoothQuantConfig(alpha=0.5)
q_model = quantize(model=float_model, quant_config=sq_config)


# Option 2 (Recommended for v2.5)
# usage 1
from neural_compressor.torch.quantization import SmoothConfig, StaticQuantConfig, quantize
sq_config = SmoothConfig(alpha=0.5)
static_config = StaticQuantConfig(w_sym=False, w_algo="minmax", white_list=["linear1"])
q_model = quantize(model=float_model, quant_config=[sq_config, static_config])

# usage 2
from neural_compressor.torch.quantization import quantize, get_default_smooth_quant_config
q_model = quantize(model=float_model, quant_config=get_default_smooth_quant_config())


# Option 3 (Recommended for further)
from neural_compressor.torch import SmoothConfig, optimize
sq_config = SmoothConfig(alpha=0.5)
# optimize: float model -> float model
optimized_model = optimize(model=float_model, optimize_config=composite_config)

from neural_compressor.torch import StaticQuantConfig, quantize
# quantize: float model -> quantized model
static_config = StaticQuantConfig(w_sym=False, w_algo="minmax", white_list=["linear1"])
q_model = quantize(model=optimized_model, quant_config=static_config)
```

**Comment by yiliu30** [comment]
### Unify and extend `register_algo`
- [ ] Unify the `register_algo`
- [ ] Clean code, import from common

```python
def register_algo(name: str):

```

```python
DEFAULT_BACKEND = "default_backend"
def register_algo(name: str , backend: str =DEFAULT_BACKEND):
    ...

```

**Comment by yiliu30** [comment]
![inc3_arch](https://github.com/intel/neural-compressor/assets/106061964/72070488-a0f5-42b1-9992-3addb11b9118)
